src/util/file_watch.c

```diff
@@ -0,0 +1,308 @@
+        if (ret != EOK) {
+            DEBUG(SSSDBG_TRACE_LIBS, "Cannot create poll timer\n");
+            return ret;
+        }
+    }
+
+    return EOK;
+}
+
+
+static int watch_file(TALLOC_CTX *mem_ctx,
+                      struct file_watch_ctx *fw_ctx)
+{
+    int ret = EOK;
+    bool use_inotify;
+
+    use_inotify = fw_ctx->use_inotify;
+    if (use_inotify) {
+        ret = try_inotify(fw_ctx);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_TRACE_LIBS, "Falling back to polling\n");
+            use_inotify = false;
+        }
+    }
+
+    if (!use_inotify) {
+        ret = fw_watch_file_fallback(mem_ctx, fw_ctx);
+    }
+
+    return ret;
+}
+
+
+static void missing_file(struct tevent_context *ev,
+                                struct tevent_timer *te,
+                                struct timeval tv, void *data)
+{
+    int ret;
+    struct file_watch_ctx *fw_ctx = talloc_get_type(data, struct file_watch_ctx);
+
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == EOK) {
+        fw_ctx->cb(fw_ctx->filename, fw_ctx->cb_arg);
+    } else if (ret == ENOENT) {
+        DEBUG(SSSDBG_TRACE_LIBS,
+              "%s missing. Waiting for it to appear.\n",
+              fw_ctx->filename);
+        tv = tevent_timeval_current_ofs(MISSING_FILE_POLL_TIME, 0);
+        te = tevent_add_timer(fw_ctx->ev, fw_ctx, tv, missing_file, fw_ctx);
+        if (te == NULL) {
+            DEBUG(SSSDBG_IMPORTANT_INFO,
+                  "tevent_add_timer failed. %s will be ignored.\n",
+                  fw_ctx->filename);
+        }
+    } else {
+        DEBUG(SSSDBG_IMPORTANT_INFO,
+              "watch_file failed. %s will be ignored: [%i] %s\n",
+              fw_ctx->filename, ret, sss_strerror(ret));
+    }
+}
+
+
+struct file_watch_ctx *fw_watch_file(TALLOC_CTX *mem_ctx,
+                                     struct tevent_context *ev,
+                                     const char *filename,
+                                     bool use_inotify,
+                                     fw_callback cb,
+                                     void *cb_arg)
+{
+    int ret;
+    struct timeval tv;
+    struct tevent_timer *te;
+    struct file_watch_ctx *fw_ctx;
+
+    if (ev == NULL || filename == NULL || cb == NULL) {
+        DEBUG(SSSDBG_TRACE_LIBS, "Invalid parameter\n");
+        return NULL;
+    }
+
+    fw_ctx = talloc_zero(mem_ctx, struct file_watch_ctx);
+    if (fw_ctx == NULL) {
+        DEBUG(SSSDBG_IMPORTANT_INFO, "Failed to allocate the context\n");
+        return NULL;
+    }
+
+    fw_ctx->ev = ev;
+    fw_ctx->use_inotify = use_inotify;
+    fw_ctx->cb = cb;
+    fw_ctx->cb_arg = cb_arg;
+    fw_ctx->filename = talloc_strdup(fw_ctx, filename);
+    if (fw_ctx->filename == NULL) {
+        DEBUG(SSSDBG_IMPORTANT_INFO, "talloc_strdup() failed\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    /* Watch for changes to the requested file */
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == ENOENT) {
```

Comment:
Apologies for ignorant questions, Is this code, `fw_watch_file()`, only used in SSSD (at the moment) for /etc/resolv.conf

Is it correct we will always loop and continuously wait for the file to exist when a non-existing file is requested to be watched?

Reply:
> Apologies for ignorant questions, Is this code, `fw_watch_file()`, only used in SSSD (at the moment) to watch changes to `/etc/resolv.conf` ?

Yes, it is currently only used by SSSD to monitor `/etc/resolv.conf` in the backends. It is part of `libsss_util.so` and with time we could use it to monitor other files.

> Is it correct we will always loop and continuously wait for the file to exist when a non-existing file is requested to be watched?

The initial goal was to monitor the file from the backends instead of the monitor and the existing code was moved from the monitor to the library. We could consider parameterize this looping, but I think it is better to do that when this is actually needed.